contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+      let acct = await api.db.findOne('accounts', { account: api.sender });
+      if (acct === null) {
+        acct = {
+          account: api.sender,
+          weights: [],
+        };
+        acct = await api.db.insert('accounts', acct);
+      }
+
+      const approval = await api.db.findOne('approvals', { from: api.sender, to: candidate._id });
+      if (api.assert(approval !== null, 'you have not approved this candidate')) {
+        await api.db.remove('approvals', approval);
+
+        const balance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: inst.voteToken });
+        let approvalWeight = 0;
+        if (balance && balance.stake) {
+          approvalWeight = balance.stake;
+        }
+        if (balance && balance.delegationsIn) {
+          approvalWeight = api.BigNumber(approvalWeight)
+            .plus(balance.delegationsIn)
+            .toFixed(voteTokenObj.precision, api.BigNumber.ROUND_HALF_UP);
+        }
+        const wIndex = acct.weights.findIndex(x => x.symbol === inst.voteToken);
+        if (wIndex !== -1) {
+          acct.weights[wIndex].weight = approvalWeight;
+        } else {
+          acct.weights.push({ symbol: inst.voteToken, weight: approvalWeight });
+        }
+        await api.db.update('accounts', acct);
+        await updateCandidateWeight(candidate._id, api.BigNumber(approvalWeight).negated());
+        api.emit('disapproveCandidate', { id: candidate._id });
+      }
+    }
+  }
+};
+
+actions.updateCandidateApprovals = async (payload) => {
+  const { account, token, callingContractInfo } = payload;
+
+  if (callingContractInfo === undefined) return;
+  if (callingContractInfo.name !== 'tokens') return;
+
+  const acct = await api.db.findOne('accounts', { account });
+  if (acct !== null) {
+    const params = await api.db.findOne('params', {});
+
+    // only update existing weights
+    const wIndex = acct.weights.findIndex(x => x.symbol === token.symbol);
+    if (wIndex !== -1) {
+      // calculate approval weight of the account
+      const balance = await api.db.findOneInTable('tokens', 'balances', { account, symbol: token.symbol });
+      let approvalWeight = 0;
+      if (balance && balance.stake) {
+        approvalWeight = balance.stake;
+      }
+
+      if (balance && balance.delegationsIn) {
+        approvalWeight = api.BigNumber(approvalWeight)
+          .plus(balance.delegationsIn)
+          .toFixed(token.precision, api.BigNumber.ROUND_HALF_UP);
+      }
+
+      let oldApprovalWeight = 0;
+      oldApprovalWeight = acct.weights[wIndex].weight;
+      acct.weights[wIndex].weight = approvalWeight;
+
+      const deltaApprovalWeight = api.BigNumber(approvalWeight)
+        .minus(oldApprovalWeight)
+        .dp(token.precision, api.BigNumber.ROUND_HALF_UP);
+
+      if (!api.BigNumber(deltaApprovalWeight).eq(0)) {
+        await api.db.update('accounts', acct);
+        const approvals = await api.db.find('approvals',
```

Comment:
I think we should query by voteToken here, to limit the scope for a given token stake change